AraEvent/AraEventCalibrator.cxx

```diff
@@ -1458,11 +1491,159 @@ Bool_t AraEventCalibrator::TrimFirstBlock(UsefulAtriStationEvent *theEvent, std:
+                    const double thisFreq = i*dFreq;
+                    if(thisFreq <= lowPassFreq) { // only use low frequency data
+                        angFreq.push_back(2.*M_PI*thisFreq);
+
+                        specMag.push_back(spec[i].getAbs());
+                        specPhase.push_back(spec[i].getPhase());
+                    }
+                }
+                
+                // calculate the "model" whose parameters we solve for
+                const int nAngFreq = int(angFreq.size()); // number of frequencies in low frequency data
+                std::vector< std::vector<double> > constMag(num_block, std::vector<double>(nAngFreq));
+                std::vector< std::vector<double> > constPhase(num_block, std::vector<double>(nAngFreq));
+                for(int b = 0; b < num_block; ++b) {
+                    const double blockWidth = dT[b];
+                    const double blockStartShift = blockStarts[b] - blockStarts[0];
+                    for(int i = 0; i < nAngFreq; ++i) {
+                        const double thisAngFreq = angFreq[i];
+
+                        if(thisAngFreq == 0)
+                            constMag[b][i] = blockWidth/dt;
+                        else
+                            constMag[b][i] = TMath::Sin(blockWidth*thisAngFreq/2.)/TMath::Sin(dt*thisAngFreq/2.); 
+                        constPhase[b][i] = -thisAngFreq*(blockWidth/2.+blockStartShift);
+                    }
+                }
+
+                // get block offsets
+                // we do this by minimizing the squared error between the low passed spectrum and the block offsets
+                // this can be done by solving a matrix equation, so we start by setting that up
+                // note that the squared error has been simplified here so it can be written directly in terms of real numbers
+                TMatrixDSym M(num_block);
+                TVectorD b(num_block);
+                for(int b1 = 0; b1 < num_block; ++b1) {
+                    
+                    for(int b2 = 0; b2 <= b1; ++b2) {
+                        
+                        std::vector<double>& constMag2 = constMag[b2];
+                        std::vector<double>& constPhase2 = constPhase[b2];
+                        double valM = 0.;
+                        for(int i = 0; i < nAngFreq; ++i)
+                            valM += constMag[b1][i] * constMag[b2][i] * TMath::Cos(constPhase[b1][i] - constPhase[b2][i]);
+                        M(b1, b2) = valM;
+                    }
+                    
+                    double valb = 0.;
+                    for(int i = 0; i < nAngFreq; ++i)
+                        valb += specMag[i] * constMag[b1][i] * TMath::Cos(specPhase[i] - constPhase[b1][i]);
+                    b(b1) = valb;
+                }               
+
+                // now we solve the equation M*A = b, where A is a vector of block offset amplitudes
+                TDecompSVD svd(M);
```

Comment:
how much of this can be amortized / cached?

Reply:
In principle, I think the matrix is entirely cache-able if the block durations and start times relative to the beginning of the first block were all the same. But I'm not sure this can be assumed on ARA in general? Maybe it's true for all the channels on a single DDA, though. I can check.